### DISCUSSION TITLE
Loading OpenType Features from UFO

### DISCUSSION THREAD

**Discussion by bensapirstein** [coldtype/coldtype#143] [Q&A]
I'm currently working on the development of a variable font and using Coldtype to debug the animations. While working with the raw UFO file, I encountered an issue when trying to use certain OpenType features, such as Arabic text "ملأك" or special characters like "à". The problem results in a program crash and provides the following traceback:

```arduino
Traceback (most recent call last):
  File "/Users/bensa/coldtype/Wasim/venv/lib/python3.9/site-packages/coldtype/renderer/__init__.py", line 535, in _single_thread_render
    result = render.normalize_result(render.run(rp, self.state))
  File "/Users/bensa/coldtype/Wasim/venv/lib/python3.9/site-packages/coldtype/renderable/renderable.py", line 346, in run
    res = render_pass.fn(*render_pass.args)
  ...
  File "/Users/bensa/coldtype/Wasim/venv/lib/python3.9/site-packages/fontTools/pens/transformPen.py", line 55, in addComponent
    self._outPen.addComponent(glyphName, transformation)
AttributeError: 'P' object has no attribute 'addComponent'
```

This error occurs when attempting to render simple text using the following Python code:

```python
from coldtype import *
from coldtype.fx.skia import phototype

fnt = Font("IBM Plex/IBM plex master IBM-Plex-Sans-Arabic-sources_masters/IBM Plex Sans Arabic-Regular.ufo")
txt = "ملأك"

@renderable((1080, 1080))
def text(r):
    return (StSt(txt, fnt, 400).pen()
            .align(r)
            .f(1)
            .ch(phototype(r, blur=3, cut=90, cutw=30, fill=bw(0)))
            )

```

**Comment by stenson** [comment]
Ah! Yeah I've been meaning to figure out how to support `addComponent` in cases like this — basically when load a ufo, it does a minimal (i.e. fast-as-possible) compilation of the ufo source into a ttf, but (I think) that compilation doesn't resolve components, so the Coldtype typesetting code needs to account for that — I'll make an issue for this and try to figure it out soon.

**Reply by stenson** [reply]
In the meantime, it looks like it might work if you use a .designspace to load the .ufo's indirectly? Just tried it with "IBM Plex Sans Arabic.designspace" and it loaded the componentized glyphs correctly.

**Comment by stenson** [comment]
I don't have an exact fix for the `addComponent` error, but the latest Coldtype (0.10.14) does introduce a `Font.Fontmake` method that does a full compilation of a ufo (or designspace) file to a tmp .ttf (into an in-memory cache managed by Coldtype). So for your example above, you could change the code to:

```python
from coldtype import *
from coldtype.fx.skia import phototype

fnt = Font.Fontmake("IBM Plex/IBM plex master IBM-Plex-Sans-Arabic-sources_masters/IBM Plex Sans Arabic-Regular.ufo")
txt = "ملأك"

@renderable((1080, 1080))
def text(r):
    return (StSt(txt, fnt, 400).pen()
            .align(r)
            .f(1)
            .ch(phototype(r, blur=3, cut=90, cutw=30, fill=bw(0)))
            )
```

and it should be work properly (if you `pip install fontmake` in your virtual environment). It will be a lot slower than the previous UFO compilation, but it should also be a lot more accurate.

**Reply by bensapirstein** [reply]
Thanks Rob, this works at compiling the font and the substitution features :)
So far I avoided compiling fonts and preferred working with the raw UFO since I'm using stroke-based representations that rely on open contours. I remember you suggested I'd try the --keep-overlaps flag but it has no effect, and the resulting font had closed contours. 

Moreover, I would like to add the font/sources to the watch list so that when I make changes to the sources they will be compiled and updated automatically in the animation, how could that be achieved? I tried adding them to the watch list but it doesn't recompile the var font.

```python
from coldtype import *
from coldtype.fx.skia import phototype

fnt_ar = Font("Wasim/sources/Wasim-Ar.ufo")
fnt_hb = Font("Wasim/sources/Wasim-Heb.ufo")

fnt = Font.Fontmake("Wasim/Wasim.designspace")

l = 'مم'

@animation(timeline=Timeline(180, 30), bg=bw(1), watch=[fnt_ar.path, fnt_hb.path])
def anim(f):
    return (StSt(l, fnt, 1000, wght=f.e("eeio"))
            .explode()
            .outline(1)
            .castshadow(40, 80)
            .align(f.a.r)
            .f(1)
            .ch(phototype(f.a.r, blur=3, cut=90, cutw=30, fill=bw(0)))
            )
```


https://github.com/coldtype/coldtype/assets/14977606/a1a15eaa-1d3e-40ca-b39a-6b8061106e70

**Reply by stenson** [reply]
Is there a chance you could send me a couple example ufos that you're working with? Feel like I'd be able to get a better handle on it if I had the real ufos to work with while debugging.

**Reply by stenson** [reply]
I _think_ this code should do what you’re looking for, but I had to update Font.Fontmake for it to work. I haven't released that to the official Coldtype yet, so you'll need to run something like `pip install git+https://github.com/coldtype/coldtype` to get it (and you might need to `pip uninstall coldtype` before you do that to make sure you've got the latest directly from git)

```python
from coldtype import *
from coldtype.fx.skia import phototype

fnt_ar = Font("~/Downloads/Wasim/sources/Wasim-Ar.ufo")
fnt_hb = Font("~/Downloads/Wasim/sources/Wasim-Heb.ufo")

fnt = Font.Fontmake("~/Downloads/Wasim/Wasim.designspace")

l = 'مم'

def unclose(p:P):
    p._val.value = p._val.value[:-1]
    p.endPath()
    return p

@animation(timeline=Timeline(180, 30), bg=bw(1), watch=[fnt_ar.path, fnt_hb.path])
def anim(f):
    return (StSt(l, fnt, 1000, wght=f.e("eeio"))
        .mapv(unclose)
        .explode()
        .fssw(-1, 0, 1)
        .outline(1)
        .castshadow(40, 80)
        .align(f.a.r)
        .f(1)
        .ch(phototype(f.a.r, blur=3, cut=90, cutw=30, fill=bw(0)))
        )
```

**Comment by bensapirstein** [comment]
Thanks Rob this is perfect for my purpose!

What about automatically updating the _fnt_ variable with a recompiled font when one of its sources _fntAr_ or _fntHb_ is modified?

**Reply by stenson** [reply]
Should be working already (or worked on my setup when I modified one of the ufos)

**Reply by bensapirstein** [reply]
Had to uninstall coldtype first, now it works perfectly. Thousand thanks 🙏🙏
